packages/@aws-cdk/aws-ecs/README.md

```diff
@@ -1,2 +1,142 @@
-## The CDK Construct Library for AWS Elastic Container Service (ECS)
-This module is part of the [AWS Cloud Development Kit](https://github.com/awslabs/aws-cdk) project.
+## AWS Elastic Container Service (ECS) Construct Library
+
+This package contains constructs for working with **AWS Elastic Container
+Service** (ECS). The simplest example of using this library looks like this:
+
+```ts
+// Create an ECS cluster (backed by an AutoScaling group)
+const cluster = new ecs.Ec2Cluster(this, 'Cluster', {
+  vpc,
+  size: 3,
+  instanceType: new InstanceType("t2.xlarge")
+});
+
+// Instantiate ECS Service with an automatic load balancer
+const ecsService = new ecs.LoadBalancedEc2Service(this, 'Service', {
+  cluster,
+  memoryLimitMiB: 512,
+  image: ecs.DockerHub.image("amazon/amazon-ecs-sample"),
+});
+```
+
+### Fargate vs ECS
+
+There are two sets of constructs in this library; one to run tasks on ECS and
+one to run Tasks on fargate.
+
+- Use the `Ec2Cluster`, `Ec2TaskDefinition` and `Ec2Service` constructs to
+  run tasks on EC2 instances running in your account.
+- Use the `FargateCluster`, `FargateTaskDefinition` and `FargateService`
+  constructs to run tasks on instances that are managed for you by AWS.
+
```

Comment:
From reading this, I don't understand why we need specialized classes for FargateTaskDefinition and FargateService. I am assuming I'll find out why later on, but intuitively I expected just the cluster to have a different concrete class.

The reason why this might be an issue for customers is if they want to port an ECS service/task-def from EC2 to Fargate. Ideally, if they could just change the cluster and it will Just Work, that would have been a nice capability.

Reply:
I know, and this was our original plan as well. But there are a number of differences that make this tricky:

* There is a difference in how you specify task sizes. For EC2 tasks, you must specify task memory limits on the container. For Fargate tasks, you must specify instance sizes on the TaskDefinition.
* EC2 task definitions allow different placement constraints.
* Fargate task definitions do not allow configuring networkmode.

After discussion we decided that task migration wasn't that likely, so it was okay if it didn't work out of the box.

Reply:
Can we document these differences in the README?

Still feels like we are missing something... For example, if a 3rd party library wants to vend a `TaskDefinition` for reuse, they will have to either vend two (one for Fargate and one for EC2) or users will not be able to use them interchangeably. 

Maybe what we can do is define a `TaskDefinition` which is the union of all options, so I can define a task definition that can be used across EC2/Fargate. Does that make any sense?

Reply:
I don't think it should be broken up--we don't break it up in the console even. One of the things that we talk about is how you can easily migrate from EC2 to Fargate and vice versa. Also if you use awsvpc you can easily create a task definition that works for both. There isn't any field in a task definition that makes it Fargate or EC2 specific--just the values or IAM roles you give--but both can take the same things.

Reply:
@rix0rrr What do you mean you specify instance sizes in a task definition? You can only have task/container mem/cpu.

Reply:
Console is changeable :)  I do not think the way to create a "Fargate" task definition is intuitive, especially when customers come looking to use "Fargate" and have to decipher how to create a "Fargate-compatible" task definition in ECS with a combination of attributes and set the compatibility flag.  I want them to have first-class Fargate objects all the way if they come looking to use Fargate.  It sounds more like we *also* need more granular object types for sophisticated customers who already understand some of these concepts.

I think it is still worth having a top-level Fargate task definition object, even if we do something like have: TaskDefinition (set whatever you want), FargateTaskDefinition, and EC2TaskDefinition.

Reply:
@tiffanyfay In re: "There isn't any field in a task definition that makes it Fargate or EC2 specific" -- not quite true.  There are certain fields on TaskDefinition today that are not applicable to Fargate at both the top-level TaskDefinition level as well as certain ContainerDefinition fields (some of which are coupled, in the case of named volumes), and vice versa (e.g. task size). Setting up AwsVpc configuration is also non-trivial, as that is tied to specific network resources in a way that is not the case for non-Fargate task definitions.

In other words, I agree with @clareliguori that creating a Fargate task definition is not intuitive, and that the CDK provides an opportunity to streamline workflows that perhaps are not available in the console. While we can certainly benefit from decisions the console has already made, I think we have the opportunity to add value by doing things differently in certain cases from existing tooling.

In re: having one cluster, I addressed that [here](https://github.com/awslabs/aws-cdk/pull/1041#discussion_r229894358) -- if you are interested in the exact numbers, I can show you but the number of heterogenous clusters is extremely low (less than a fractional percentage of total clusters, and also extremely low percentage of Fargate customers).

Reply:
Also, by "instance size", Rico was referring to task size, which can be thought of as just instance size since that is the actual resource the scheduler is considering when given mem/cpu limits. @tiffanyfay

Reply:
Begin able to add instances also sounds like a useful feature

Reply:
If you look at the current implementation in `Ec2Cluster`, that is currently what we are doing already, except via an ASG. @rix0rrr and I had already discussed this following [this other comment](https://github.com/awslabs/aws-cdk/pull/1041#discussion_r229560140) regarding configurability of ASGs. The reason why we wanted to keep this separate was to hide any properties having to do with instances from the Fargate namespace.

I'm not sure what you mean by `size` in your `addInstances` example -- the `size` field on the `Ec2Cluster` construct currently refers to the size of the cluster, e.g. number of instances. I assume you mean count?

Finally, Rico and I had also already discussed that userData is something that you should be able to add separately -- I believe it was still in the context of an ASG.

In re: adding instances separately, do we actually have data supporting instances existing in clusters outside of an ASG? Both the console and the ecs-cli currently only provision capacity through ASGs today, which is why we modeled the `Ec2Cluster` construct this way. I'm not against the idea of adding an API that adds individual capacity, but it seems a little strange to want instances that isn't part of an ASG, but perhaps @tiffanyfay you have a concrete customer use case?

Reply:
Yea, saw that and think that the ASG should be moved into the `addInstances` instead. So since ECS has a cluster which has both EC2 and Fargate vs them being individual clusters, I think there should be just one `Cluster` construct.

My thought of `addInstances` (or maybe it should be addInstanceASG or something) is not to scale existing types, but to add new types. Examples being:
-Create cluster with size: 3 type: t2.micro and I want to add 4 t2.medium instances in an ASG.
-Create cluster without instances (e.g. Fargate) and add 4 t2.medium instances in an ASG.

Basically I want to easily be able to have different types of instances. So actually it would need to only allow one asg in order to return something you can scale individually.

Basically the RegisterContainerInstance API plus creating the asg/instances for you.

Oh, yes, count-- I had just copy pasted.

Would it be able to be added during instance/asg creation or only after the fact and updating?

Reply:
The `RegisterContainerInstance` is an internal API only used by the agent; it's not externally exposed to customers IIRC. I think having an `addCapacity` API isn't a bad idea, as long as it's well-defined and aligned with best practices. Currently, we have plans to allow instance auto-scaling (which @rix0rrr is working on). 

For me I am still not convinced there is good justification for not separating the cluster types. As previously mentioned, there is an extremely low number of customers today who run both kinds of tasks/services in the same cluster -- we originally supported it to ease migration, but in today's world where both services are available as options it seems like we should be supporting a more streamlined workflow. Do you happen to have concrete customer use cases for what kind of system is running both kinds of services?